src/v/resource_mgmt/memory_sampling.cc

```diff
@@ -0,0 +1,224 @@
+                                   format_buf.size(),
+                                   "{}",
+                                   allocation_sites[i])
+                                   .size;
+
+            writer(std::string_view(format_buf.data(), bytes_written));
+        }
+    };
+}
+
+/// We want to print the top-n allocation sites on OOM
+/// Set this up and make sure we don't allocate extra at that point
+void setup_additional_oom_diagnostics() {
+    seastar::memory::set_additional_diagnostics_producer(
+      memory_sampling::get_oom_diagnostics_callback());
+}
+
+void memory_sampling::start_low_available_memory_logging() {
+    // We want some periodic logging "on the way" to OOM. At the same time we
+    // don't want to spam the logs. Hence, we periodically look at the available
+    // memory low watermark (this is without the batch cache). If we see that we
+    // have crossed the 10% and 20% marks we log the allocation sites. We stop
+    // afterwards.
+    //
+    // This logic could also be integrated into batch_cache::reclaim (so that we
+    // actually run when memory is tight) but that is quite critical code so
+    // better not to add anything not needed there.
+    size_t first_log_limit = _first_log_limit_percent
+                             * seastar::memory::stats().total_memory();
+    size_t second_log_limit = _second_log_limit_percent
+                              * seastar::memory::stats().total_memory();
+    size_t next_log_limit = first_log_limit;
+    size_t last_seen_watermark = std::numeric_limits<size_t>::max();
+
+    _timer.set_callback([this,
+                         last_seen_watermark,
+                         first_log_limit,
+                         second_log_limit,
+                         next_log_limit]() mutable {
+        auto current_low_water_mark
+          = resources::available_memory::local().available_low_water_mark();
+
+        if (
+          current_low_water_mark > next_log_limit
+          || current_low_water_mark >= last_seen_watermark) {
+            return;
+        }
+
+        last_seen_watermark = current_low_water_mark;
+
+        auto allocation_sites = ss::memory::sampled_memory_profile();
+        const size_t top_n = std::min(size_t(5), allocation_sites.size());
+        top_n_allocation_sites(allocation_sites, top_n);
+
+        vlog(
+          _logger.info,
+          "{} {}",
+          diagnostics_header(),
+          fmt::join(
+            allocation_sites.begin(), allocation_sites.begin() + top_n, "|"));
+
+        if (next_log_limit == first_log_limit) {
+            next_log_limit = second_log_limit;
+        } else {
+            _timer.cancel();
+            return;
+        }
+    });
+
+    _timer.arm_periodic(
+      std::chrono::seconds(_low_watermark_check_frequency_seconds));
+}
+
+memory_sampling::memory_sampling(ss::logger& logger)
+  : _logger(logger)
+  , _first_log_limit_percent(0.2)
```

Comment:
This is 20% not 0.2% right? In that case we can name it like "fraction" instead of percent.